File: MathProg/math.js
// find a square root of given number:
// num = parseInt(prompt("Enter number: "))
// console.log(`Sqaure root of ${num} = ${Math.sqrt(num)}`)
// ------------------------------------------------
// sqrt = (sqr) => sqr ** 0.5
// num_sqrt = 56 //parseInt(prompt("Enter number: "))

// cbrt = (cbr) => cbr ** (1 / 3)
// num_cbrt = 64//parseInt(prompt("Enter number: "))

// power = (base, exponent) => base ** exponent
// base = 2 //parseInt(prompt("Enter base: "))
// exponent = 3 //parseInt(prompt("Enter exponent: "))


// console.log(`Sqaure root of ${num_sqrt} = ${sqrt(num_sqrt)}`)
// console.log(`Cube root of ${num_cbrt} = ${cbrt(num_cbrt)}`)
// console.log(`The ${base} the power of ${exponent} = ${power(base, exponent)}`)
// -------------------------------------------
// function cbrt(x) {
//     return x >= 0 ? x ** (1 / 3) : NaN;
// }

// console.log(cbrt(125));  // Output: 3

// function cbrt(x) {
//     if (x < 0) return -cbrt(-x);
//     if (x === 0) return 0;
//     let y = x;
//     let e = 0.000000001;
//     while (Math.abs(y * y * y - x) > e) {
//         y = (2 * y + x / (y * y)) / 3;
//     }
//     return y;
// }

// console.log(cbrt(-64));  // Output: 4
// console.log(cbrt(125));  // Output: 5

// -----------------------------------
// {

//     cbrt = (x) => {
//         if (x < 0) return -cbrt(-x)
//         if (x === 0) return 0
//         let low = 0
//         let high = x
//         let e = 0.000000001
//         while ((high - low) > e) {
//             let mid = (high + low) / 2
//             if ((mid * mid * mid) <= x) {
//                 low = mid
//             } else {
//                 high = mid
//             }
//         }
//         return low
//     }
//     console.log(cbrt(64))
//     console.log(cbrt(125))
// }

{
    function sqrt(x) {
        if (x < 0) throw new Error("Cannor take square root of negative value")
        if (x === 0) return 0
        return Math.sqrt(x)
    }

    console.log(sqrt(16))
    console.log(Math.sqrt(25))
}

{
    function cbrt(x) {
        if (x < 0) return -Math.cbrt(-x);
        if (x === 0) return 0;
        return Math.cbrt(x);
    }

    console.log(cbrt(64));  // Output: 4
    console.log(cbrt(125));  // Output: 5
}
{
    function cbrt(x) {
        if (x < 0) return -Math.cbrt(-x);
        if (x === 0) return 0;
        return Math.cbrt(x);
    }

    console.log(cbrt(64));  // Output: 4
    console.log(cbrt(125));  // Output: 5
}

File: MathProg/math.cs
// public static double SquareRoot(double number)
// {
//     double x = number;
//     double y = 1;
//     double e = 0.000001; // precision

//     while (x - y > e)
//     {
//         x = (x + y) / 2;
//         y = number / x;
//     }

//     return x;
// }
// public static double CubeRoot(double number)
// {
//     double x = number;
//     double y = 1;
//     double e = 0.000001; // precision

//     while (Math.Abs(x - y) > e)
//     {
//         y = ((2 * x) + (number / (x * x))) / 3;
//         x = y;
//     }

//     return x;
// }
// public static double Power(double baseNumber, double exponent)
// {
//     double result = 1.0;

//     for (int i = 0; i < exponent; i++)
//     {
//         result *= baseNumber;
//     }

//     return result;
// }

// double sqrtResult = SquareRoot(16.0); // returns 4.0
// double cbrtResult = CubeRoot(27.0); // returns 3.0
// double powerResult = Power(2.0, 3.0); // returns 8.0


using System;
class Program
{
    // calculate the square root:
    public static double SquareRoot(double number)
    {
        return Math.Sqrt(number);
    }
    // calculate the cube root:
    public static double CubeRoot(double number)
    {
        return Math.Pow(number, 1.0 / 3.0);
    }
    // calculate the power of a number:
    public static double Power(double baseNumber, double exponent)
    {
        return Math.Pow(baseNumber, exponent);
    }
    static void Main(string[] args)
    {
        double sqrtResult, cbrtResult, baseNumber, exponent;

        Console.Write("Enter the number for Sqaure Root: ");
        sqrtResult = Convert.ToDouble(Console.ReadLine());
        Console.WriteLine($"Square Root of {sqrtResult} = {SquareRoot(sqrtResult)}");

        Console.Write("\nEnter the number for Cube Root: ");
        cbrtResult = Convert.ToDouble(Console.ReadLine());
        Console.WriteLine($"The Cube Root of {cbrtResult} = {CubeRoot(cbrtResult)}");

        Console.Write("\nEnter the base: ");
        baseNumber = Convert.ToDouble(Console.ReadLine());
        Console.Write("Enter the exponent: ");
        exponent = Convert.ToDouble(Console.ReadLine());
        Console.WriteLine($"{baseNumber} raised to the power of {exponent} = {Power(baseNumber, exponent)}");

    }
}